File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/BasalBodyTemperatureRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.BasalBodyTemperatureRecord
import androidx.health.connect.client.records.BodyTemperatureMeasurementLocation
import androidx.health.connect.client.units.Temperature
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toTemperatureDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BasalBodyTemperatureMeasurementLocationDto
import com.phamtunglam.health_connector_hc_android.pigeon.BasalBodyTemperatureRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [BasalBodyTemperatureRecord] object to a [BasalBodyTemperatureRecordDto].
 */
internal fun BasalBodyTemperatureRecord.toDto(): BasalBodyTemperatureRecordDto =
    BasalBodyTemperatureRecordDto(
        id = metadata.id,
        time = time.toEpochMilli(),
        zoneOffsetSeconds = zoneOffset?.totalSeconds?.toLong(),
        metadata = metadata.toDto(),
        temperature = temperature.inCelsius.toTemperatureDto(),
        measurementLocation = measurementLocation.toBasalBodyTemperatureMeasurementLocationDto(),
    )

/**
 * Converts a [BasalBodyTemperatureRecordDto] to a Health Connect [BasalBodyTemperatureRecord] object.
 */
internal fun BasalBodyTemperatureRecordDto.toHealthConnect(): BasalBodyTemperatureRecord =
    BasalBodyTemperatureRecord(
        temperature = Temperature.celsius(temperature.value),
        measurementLocation = measurementLocation.toHealthConnect(),
        time = Instant.ofEpochMilli(time),
        zoneOffset = zoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )

/**
 * Converts a Health Connect [BasalBodyTemperatureRecord] measurement location to a
 * [BasalBodyTemperatureMeasurementLocationDto].
 */
internal fun Int.toBasalBodyTemperatureMeasurementLocationDto():
    BasalBodyTemperatureMeasurementLocationDto =
    when (this) {
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_ARMPIT ->
            BasalBodyTemperatureMeasurementLocationDto.ARMPIT

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_EAR ->
            BasalBodyTemperatureMeasurementLocationDto.EAR

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FINGER ->
            BasalBodyTemperatureMeasurementLocationDto.FINGER

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FOREHEAD ->
            BasalBodyTemperatureMeasurementLocationDto.FOREHEAD

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_MOUTH ->
            BasalBodyTemperatureMeasurementLocationDto.MOUTH

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_RECTUM ->
            BasalBodyTemperatureMeasurementLocationDto.RECTUM

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TEMPORAL_ARTERY ->
            BasalBodyTemperatureMeasurementLocationDto.TEMPORAL_ARTERY

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TOE ->
            BasalBodyTemperatureMeasurementLocationDto.TOE

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_VAGINA ->
            BasalBodyTemperatureMeasurementLocationDto.VAGINA

        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_WRIST ->
            BasalBodyTemperatureMeasurementLocationDto.WRIST

        else -> BasalBodyTemperatureMeasurementLocationDto.UNKNOWN
    }

/**
 * Converts a [BasalBodyTemperatureMeasurementLocationDto] to a Health Connect
 * [BasalBodyTemperatureRecord] measurement location.
 */
internal fun BasalBodyTemperatureMeasurementLocationDto.toHealthConnect(): Int = when (this) {
    BasalBodyTemperatureMeasurementLocationDto.UNKNOWN ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_UNKNOWN

    BasalBodyTemperatureMeasurementLocationDto.ARMPIT ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_ARMPIT

    BasalBodyTemperatureMeasurementLocationDto.EAR ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_EAR

    BasalBodyTemperatureMeasurementLocationDto.FINGER ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FINGER

    BasalBodyTemperatureMeasurementLocationDto.FOREHEAD ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FOREHEAD

    BasalBodyTemperatureMeasurementLocationDto.MOUTH ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_MOUTH

    BasalBodyTemperatureMeasurementLocationDto.RECTUM ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_RECTUM

    BasalBodyTemperatureMeasurementLocationDto.TEMPORAL_ARTERY ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TEMPORAL_ARTERY

    BasalBodyTemperatureMeasurementLocationDto.TOE ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TOE

    BasalBodyTemperatureMeasurementLocationDto.VAGINA ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_VAGINA

    BasalBodyTemperatureMeasurementLocationDto.WRIST ->
        BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_WRIST
}

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/temperature/basal_body_temperature_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show
        BasalBodyTemperatureRecord,
        BasalBodyTemperatureMeasurementLocation,
        HealthRecordId,
        sinceV2_2_0,
        Temperature;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show
        BasalBodyTemperatureRecordDto,
        BasalBodyTemperatureMeasurementLocationDto;
import 'package:meta/meta.dart' show internal;

/// Converts [BasalBodyTemperatureRecord] to [BasalBodyTemperatureRecordDto].
@sinceV2_2_0
@internal
extension BasalBodyTemperatureRecordToDto on BasalBodyTemperatureRecord {
  BasalBodyTemperatureRecordDto toDto() {
    return BasalBodyTemperatureRecordDto(
      id: id.toDto(),
      time: time.millisecondsSinceEpoch,
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDto(),
      celsius: temperature.inCelsius,
      measurementLocation: measurementLocation.toDto(),
    );
  }
}

/// Converts [BasalBodyTemperatureRecordDto] to [BasalBodyTemperatureRecord].
@sinceV2_2_0
@internal
extension BasalBodyTemperatureRecordDtoToDomain
    on BasalBodyTemperatureRecordDto {
  BasalBodyTemperatureRecord toDomain() {
    return BasalBodyTemperatureRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      time: DateTime.fromMillisecondsSinceEpoch(time, isUtc: true),
      zoneOffsetSeconds: zoneOffsetSeconds,
      metadata: metadata.toDomain(),
      temperature: Temperature.celsius(celsius),
      measurementLocation: measurementLocation.toDomain(),
    );
  }
}

/// Converts [BasalBodyTemperatureMeasurementLocationDto] to
/// [BasalBodyTemperatureMeasurementLocation].
@sinceV2_2_0
@internal
extension BasalBodyTemperatureMeasurementLocationDtoToDomain
    on BasalBodyTemperatureMeasurementLocationDto {
  BasalBodyTemperatureMeasurementLocation toDomain() {
    return switch (this) {
      BasalBodyTemperatureMeasurementLocationDto.unknown =>
        BasalBodyTemperatureMeasurementLocation.unknown,
      BasalBodyTemperatureMeasurementLocationDto.armpit =>
        BasalBodyTemperatureMeasurementLocation.armpit,
      BasalBodyTemperatureMeasurementLocationDto.ear =>
        BasalBodyTemperatureMeasurementLocation.ear,
      BasalBodyTemperatureMeasurementLocationDto.finger =>
        BasalBodyTemperatureMeasurementLocation.finger,
      BasalBodyTemperatureMeasurementLocationDto.forehead =>
        BasalBodyTemperatureMeasurementLocation.forehead,
      BasalBodyTemperatureMeasurementLocationDto.mouth =>
        BasalBodyTemperatureMeasurementLocation.mouth,
      BasalBodyTemperatureMeasurementLocationDto.rectum =>
        BasalBodyTemperatureMeasurementLocation.rectum,
      BasalBodyTemperatureMeasurementLocationDto.temporalArtery =>
        BasalBodyTemperatureMeasurementLocation.temporalArtery,
      BasalBodyTemperatureMeasurementLocationDto.toe =>
        BasalBodyTemperatureMeasurementLocation.toe,
      BasalBodyTemperatureMeasurementLocationDto.vagina =>
        BasalBodyTemperatureMeasurementLocation.vagina,
      BasalBodyTemperatureMeasurementLocationDto.wrist =>
        BasalBodyTemperatureMeasurementLocation.wrist,
    };
  }
}

/// Converts [BasalBodyTemperatureMeasurementLocation] to
/// [BasalBodyTemperatureMeasurementLocationDto].
@sinceV2_2_0
@internal
extension BasalBodyTemperatureMeasurementLocationToDto
    on BasalBodyTemperatureMeasurementLocation {
  BasalBodyTemperatureMeasurementLocationDto toDto() {
    return switch (this) {
      BasalBodyTemperatureMeasurementLocation.unknown =>
        BasalBodyTemperatureMeasurementLocationDto.unknown,
      BasalBodyTemperatureMeasurementLocation.armpit =>
        BasalBodyTemperatureMeasurementLocationDto.armpit,
      BasalBodyTemperatureMeasurementLocation.ear =>
        BasalBodyTemperatureMeasurementLocationDto.ear,
      BasalBodyTemperatureMeasurementLocation.finger =>
        BasalBodyTemperatureMeasurementLocationDto.finger,
      BasalBodyTemperatureMeasurementLocation.forehead =>
        BasalBodyTemperatureMeasurementLocationDto.forehead,
      BasalBodyTemperatureMeasurementLocation.mouth =>
        BasalBodyTemperatureMeasurementLocationDto.mouth,
      BasalBodyTemperatureMeasurementLocation.rectum =>
        BasalBodyTemperatureMeasurementLocationDto.rectum,
      BasalBodyTemperatureMeasurementLocation.temporalArtery =>
        BasalBodyTemperatureMeasurementLocationDto.temporalArtery,
      BasalBodyTemperatureMeasurementLocation.toe =>
        BasalBodyTemperatureMeasurementLocationDto.toe,
      BasalBodyTemperatureMeasurementLocation.vagina =>
        BasalBodyTemperatureMeasurementLocationDto.vagina,
      BasalBodyTemperatureMeasurementLocation.wrist =>
        BasalBodyTemperatureMeasurementLocationDto.wrist,
    };
  }
}
